lib/api/apiUtils/authorization/permissionChecks.js

```diff
@@ -0,0 +1,302 @@
+
+        if (principalMatch && actionMatch && s.Effect === 'Deny') {
+            explicitDeny = true;
+        }
+        if (principalMatch && actionMatch && s.Effect === 'Allow') {
+            allowed = true;
+        }
+        // eslint-disable-next-line no-param-reassign
+        policy.Statement = policy.Statement.splice(1);
+    }
+    if (explicitDeny) {
+        return 'explicitDeny';
+    }
+    return allowed;
+}
+
+function isBucketAuthorized(bucket, requestType, canonicalID, arn) {
+    // Check to see if user is authorized to perform a
+    // particular action on bucket based on ACLs.
+    // TODO: Add IAM checks and bucket policy checks.
+    if (bucket.getOwner() === canonicalID) {
+        return true;
+    } else if (constants.bucketOwnerActions.includes(requestType)) {
+        // only bucket owner can modify or retrieve this property of a bucket
+        return false;
+    }
+    const aclPermission = checkBucketAcls(bucket, requestType, canonicalID);
+    const bucketPolicy = bucket.getBucketPolicy();
+    if (!bucketPolicy) {
+        return aclPermission;
+    }
+    const bucketPolicyPermission = checkBucketPolicy(bucketPolicy, requestType,
+        canonicalID, arn);
+    if (bucketPolicyPermission === 'explicitDeny') {
+        return false;
+    }
+    return (aclPermission || bucketPolicyPermission);
+}
+
+function isObjAuthorized(bucket, objectMD, requestType, canonicalID, arn) {
+    const bucketOwner = bucket.getOwner();
+    if (!objectMD) {
+        return false;
+    }
+    if (objectMD['owner-id'] === canonicalID) {
+        return true;
+    }
+    // account is authorized if:
+    // - requesttype is "bucketOwnerAction" (example: for objectTagging) and
+    // - account is the bucket owner
+    if (constants.bucketOwnerActions.includes(requestType)
+    && bucketOwner === canonicalID) {
+        return true;
+    }
+    const aclPermission = checkObjectAcls(bucket, objectMD, requestType,
+        canonicalID);
+    const bucketPolicy = bucket.getBucketPolicy();
+    if (!bucketPolicy) {
+        return aclPermission;
+    }
+    const bucketPolicyPermission = checkBucketPolicy(bucketPolicy, requestType,
+        canonicalID, arn);
+    if (bucketPolicyPermission === 'explicitDeny') {
+        return false;
+    }
+    return (aclPermission || bucketPolicyPermission);
+}
+
+// the resources specified in the bucket policy should contain the bucket name
+function validatePolicyResource(bucket) {
+    const bucketName = bucket.getName();
+    const policy = bucket.getBucketPolicy();
+    const bucketArn = `arn:aws:s3:::${bucketName}`;
+
+    return policy.Statement.every(s => {
+        if (Array.isArray(s.Resource)) {
+            return s.Resource.every(r => {
+                if (r === bucketArn) {
+                    return true;
+                } else if (r.includes('/')) {
+                    const rSubs = r.split('/');
+                    return rSubs.includes(bucketArn);
```

Comment:
This check seems to succeed if any of the sub-path components in `r` *is* the bucket ARN, is this intended? I'm surprised as I would expect the sub-path components to have different semantic, and only the first one to represent the bucket ARN, the following ones could be to select objects in the bucket etc.

Reply:
Yeah, I could be more explicit and say `rSubs[0]` 👍